internal/status/gatewayclass.go

```diff
@@ -0,0 +1,39 @@
+// Copyright Project Contour Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package status
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/projectcontour/contour/internal/equality"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	gateway_v1alpha1 "sigs.k8s.io/gateway-api/apis/v1alpha1"
+)
+
+// SyncGatewayClass computes the current status of gc and updates status upon
+// any changes since last sync.
+func SyncGatewayClass(ctx context.Context, cli client.Client, gc *gateway_v1alpha1.GatewayClass, valid bool) error {
+	updated := gc.DeepCopy()
+
+	updated.Status.Conditions = mergeConditions(updated.Status.Conditions, computeGatewayClassAdmittedCondition(valid))
+
+	if equality.GatewayClassStatusChanged(gc.Status, updated.Status) {
```

Comment:
This is diverting from how we manage conditions today by creating a ConditionsAccessor (https://github.com/projectcontour/contour/blob/main/internal/status/conditions.go#L74). I think this is ok, but need to make sure we don't duplicate logic.

Reply:
Also it doesn't make sense, but what if two controllers start updating status for the Class? Can a class be shared across controllers for multiple gateways?

Reply:
@stevesloka hmm... I wonder if this is a gap in the gc status spec. Maybe gc status should be similar to route status, where conditions are specified per gateway. Otherwise we need to perform a dance when multiple contours in gateway-mode are running in the cluster. cc: @youngnick @jpeach

Reply:
I can open an upstream issue in the meantime.

Reply:
xref upstream discussion: https://kubernetes.slack.com/archives/CR0H13KGA/p1620933776270200

Reply:
@stevesloka commit `61b743d` introduces a gateway config param to enable/disable the gatewayclass controller. I think this is a simple and explicit approach that we can iterate on if needed. I was considering using gateway ns/name ownership labels but that's a bit tricky and has it's own disadvantages.

Reply:
What does this config option do? Why would I want to disable the GatewayClass controller?

Reply:
@stevesloka I initially thought that the 1st Contour would have the gc controller enabled and subsequent Contours would disable the gc controller so only 1 Contour manages gc's (since the controller string is the same for all Contours). I updated the approach in this PR so each instance of Contour will run the gc controller with a unique `gc.spec.controller` that is based on the namespace Contour is running in. This will ensure multiple instances of Contour do not fight over managing a gc.

Reply:
Feels like Contour should just not support gatewayclass. I don't think it buys us anything. Couldn't we just map to the config file or something? It's going to be a lot of work to have leader election across controllers & test/validate that.

Reply:
Let's have a chat about this, feels like we need to better map out how these pieces are going to fit together.